src/librustc_mir/interpret/terminator.rs

```diff
@@ -245,7 +245,9 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         trace!("eval_fn_call: {:#?}", fn_val);
 
         let instance = match fn_val {
-            FnVal::Instance(instance) => instance,
+            FnVal::Instance(instance) => {
+                self.subst_and_normalize_erasing_regions_in_frame(instance)
```

Comment:
This is the key point that was missing in the interpreter the whole time.

Reply:
Nice catch!

What does this enable/fix?

Reply:
Nothing really, just fixes the ICE that I was (misguidedly) trying to fix in the monomorphization/substitution code

Reply:
well, it allows us to evaluate polymorphic stuff without ICEing ;)

Reply:
Can you try if it helps [with this](https://github.com/rust-lang/rust/pull/61041#issuecomment-511113320)?